File: lachancuocgoi_kotline/app/src/main/java/com/example/lachancuocgoi/Analysis/L3/core/ResponseCache.kt
package com.example.lachancuocgoi.Analysis.L3.core

import com.example.lachancuocgoi.RiskLevel
import java.security.MessageDigest

/**
 * Simple LRU cache cho Gemini responses.
 *
 * Features:
 * - LRU eviction khi đạt maxSize
 * - Dynamic TTL based on risk level (RED=30s, ORANGE=2min, YELLOW/GREEN=5min)
 * - Thread-safe với synchronized
 *
 * @param maxSize Số lượng entries tối đa
 */
class ResponseCache<T>(
    private val maxSize: Int = 100
) {

    companion object {
        private const val TTL_RED_MS = 30_000L      // 30 seconds for RED (high risk changes fast)
        private const val TTL_ORANGE_MS = 120_000L // 2 minutes for ORANGE
        private const val TTL_DEFAULT_MS = 300_000L // 5 minutes default (GREEN/YELLOW)
    }

    private val cache = object : LinkedHashMap<String, CacheEntry<T>>(
        maxSize,
        0.75f,
        true
    ) {
        override fun removeEldestEntry(eldest: MutableMap.MutableEntry<String, CacheEntry<T>>): Boolean {
            return size > maxSize
        }
    }

    private fun getTtlForRisk(riskLevel: RiskLevel?): Long {
        return when (riskLevel) {
            RiskLevel.RED -> TTL_RED_MS
            RiskLevel.ORANGE -> TTL_ORANGE_MS
            else -> TTL_DEFAULT_MS
        }
    }

    @Synchronized
    fun get(key: String, riskLevel: RiskLevel? = null): T? {
        val hashedKey = hashKey(key)
        val entry = cache[hashedKey] ?: return null

        val ttl = getTtlForRisk(riskLevel)
        val now = System.currentTimeMillis()
        return if (now - entry.timestamp < ttl) {
            entry.value
        } else {
            cache.remove(hashedKey)
            null
        }
    }

    @Synchronized
    fun put(key: String, value: T, riskLevel: RiskLevel? = null) {
        val hashedKey = hashKey(key)
        val ttl = getTtlForRisk(riskLevel)
        cache[hashedKey] = CacheEntry(value, System.currentTimeMillis(), ttl)
    }

    private fun hashKey(text: String): String {
        val normalized = text.trim().lowercase()
        val bytes = MessageDigest.getInstance("SHA-256").digest(normalized.toByteArray())
        return bytes.joinToString("") { "%02x".format(it) }
    }

    @Synchronized
    fun getStats(): CacheStats {
        return CacheStats(
            size = cache.size,
            maxSize = maxSize
        )
    }

    @Synchronized
    fun clear() {
        cache.clear()
    }

    private data class CacheEntry<T>(
        val value: T,
        val timestamp: Long,
        val ttlMs: Long
    )
}

/**
 * Statistics về cache usage.
 */
data class CacheStats(
    val size: Int,
    val maxSize: Int
) {
    val usagePercent: Float
        get() = if (maxSize > 0) (size.toFloat() / maxSize) * 100 else 0f
}

File: lachancuoicgoi_flutter/lib/analysis/l3/core/response_cache.dart
import 'dart:collection';
import 'dart:convert';

import 'package:crypto/crypto.dart';

import '../../../core/risk_level.dart';

class ResponseCache<T> {
  ResponseCache({this.maxSize = 100});

  static const Duration _ttlRed = Duration(seconds: 30);
  static const Duration _ttlOrange = Duration(minutes: 2);
  static const Duration _ttlDefault = Duration(minutes: 5);

  final int maxSize;
  final LinkedHashMap<String, _CacheEntry<T>> _cache =
      LinkedHashMap<String, _CacheEntry<T>>();

  T? get(String key, {RiskLevel? riskLevel}) {
    final hashedKey = _hashKey(key);
    final entry = _cache.remove(hashedKey);
    if (entry == null) {
      return null;
    }
    final ttl = riskLevel == null ? entry.ttl : _ttlForRisk(riskLevel);
    if (DateTime.now().difference(entry.timestamp) >= ttl) {
      return null;
    }
    _cache[hashedKey] = entry;
    return entry.value;
  }

  void put(String key, T value, {RiskLevel? riskLevel}) {
    // Sweep expired entries trước khi insert để tránh rò rỉ
    sweepExpired();
    final hashedKey = _hashKey(key);
    _cache.remove(hashedKey);
    _cache[hashedKey] = _CacheEntry<T>(
      value: value,
      timestamp: DateTime.now(),
      ttl: _ttlForRisk(riskLevel),
    );
    while (_cache.length > maxSize) {
      _cache.remove(_cache.keys.first);
    }
  }

  /// Removes all expired entries from the cache.
  /// Returns the number of entries removed.
  int sweepExpired() {
    final now = DateTime.now();
    final keysToRemove = <String>[];
    for (final entry in _cache.entries) {
      if (now.difference(entry.value.timestamp) >= entry.value.ttl) {
        keysToRemove.add(entry.key);
      }
    }
    for (final key in keysToRemove) {
      _cache.remove(key);
    }
    return keysToRemove.length;
  }

  CacheStats getStats() {
    return CacheStats(size: _cache.length, maxSize: maxSize);
  }

  void clear() {
    _cache.clear();
  }

  Duration _ttlForRisk(RiskLevel? riskLevel) {
    return switch (riskLevel) {
      RiskLevel.red => _ttlRed,
      RiskLevel.orange => _ttlOrange,
      _ => _ttlDefault,
    };
  }

  String _hashKey(String text) {
    final normalized = text.trim().toLowerCase();
    return sha256.convert(utf8.encode(normalized)).toString();
  }
}

class CacheStats {
  const CacheStats({
    required this.size,
    required this.maxSize,
  });

  final int size;
  final int maxSize;

  double get usagePercent => maxSize > 0 ? (size / maxSize) * 100 : 0;
}

class _CacheEntry<T> {
  const _CacheEntry({
    required this.value,
    required this.timestamp,
    required this.ttl,
  });

  final T value;
  final DateTime timestamp;
  final Duration ttl;
}
